<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute progeny pathway scores and assesses significance based on permutations — progenyPerm • progeny</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Compute progeny pathway scores and assesses significance based on permutations — progenyPerm" />
<meta property="og:description" content="Compute progeny pathway scores and assesses significance based on permutations" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">progeny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ProgenySingleCell.html">Applying PROGENy on single-cell RNA-seq data</a>
    </li>
    <li>
      <a href="../articles/progenyBulk.html">PROGENy pathway signatures: Application to Bulk transcriptomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/saezlab/progeny/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/saezlab">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://www.saezlab.org">
    <span class="fas fa fas fa-university"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute progeny pathway scores and assesses significance based on permutations</h1>
    <small class="dont-index">Source: <a href='https://github.com/saezlab/progeny/blob/master/R/progenyPermutations.r'><code>R/progenyPermutations.r</code></a></small>
    <div class="hidden name"><code>progenyPerm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compute progeny pathway scores and assesses significance based on permutations</p>
    </div>

    <pre class="usage"><span class='fu'>progenyPerm</span>(
  <span class='no'>df</span>,
  <span class='no'>weight_matrix</span>,
  <span class='kw'>k</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>z_scores</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>get_nulldist</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>df</th>
      <td><p>A data.frame of n*m+1 dimension, where n is the number of omic
features to be considered and m is the number of samples/contrasts.
The first column should be the identifiers of the omic features. 
These identifiers must be coherent with the identifiers of the weight matrix.</p></td>
    </tr>
    <tr>
      <th>weight_matrix</th>
      <td><p>A progeny coefficient matrix. the first column should be
the identifiers of the omic features and should be coherent with 
the identifiers provided in df.</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>The number of permutations to be performed to generate
the null-distribution used to estimate the significance of progeny scores.  
The default value is 10000.</p></td>
    </tr>
    <tr>
      <th>z_scores</th>
      <td><p>if true, the z-scores will be returned for 
the pathway activity estimations. Else, the function returns 
a normalized z-score value between -1 and 1.</p></td>
    </tr>
    <tr>
      <th>get_nulldist</th>
      <td><p>if true, the null score distribution used for 
normalization will be returned along with the actual normalized score data 
frame.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>This function returns a list of two elements. The first element is 
a data frame of p*m+1 dimensions, where p is the number of progeny pathways,
and m is the number of samples/contrasts. Each cell represents the 
significance of a progeny pathway score for one sample/contrast. The 
significance ranges between -1 and 1. The significance is equal to x*2-1, x 
being the quantile of the progeny pathway score with respect to the null 
distribution. Thus, this significance can be interpreted as the equivalent of 
1-p.value two-sided test over an empirical distribution) with the sign 
indicating the direction of the regulation. The second element is the null 
distribution list (a null distribution is generated for each sample/contrast).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># use example gene expression matrix</span>
<span class='no'>gene_expression</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span>(<span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>,
<span class='st'>"human_input.csv"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"progeny"</span>), <span class='kw'>row.names</span> <span class='kw'>=</span> <span class='fl'>1</span>))

<span class='co'># calculate pathway activities</span>
<span class='fu'><a href='progeny.html'>progeny</a></span>(<span class='no'>gene_expression</span>, <span class='kw'>scale</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='kw'>organism</span><span class='kw'>=</span><span class='st'>"Human"</span>, <span class='kw'>top</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='kw'>perm</span><span class='kw'>=</span><span class='fl'>10000</span>)</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Michael Schubert, Alberto Valdeolivas.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


